### PR TITLE
Test Marathon upgrades 1.4.9 thru 1.6.322.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
Sometimes we are unsure if a Marathon upgrade would remove tasks.
E.g. with the task id changes in #6076 we require tests that upgrade
Marathon. This change introduces such a test infrastructure.

The test download Marathon 1.4.9, 1.5.6 and 1.6.322 and deploys an
app. It then starts on Marathon after the other and verifies that the app
is still running in the new version. I will add different types of apps
in a future release.

JIRA issues: MARATHON-8141

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/io/IO.scala [hunk 1]
```diff
@@ -75,5 +75,29 @@ object IO extends StrictLogging {
         0
     }
   }
+
+  def transfer(from: InputStream, to: OutputStream): Long = transfer(Some(from), Some(to))
+
+  // TODO(karsten): We might want to use the TarFlow
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/io/IO.scala]: So  far the tar flow is too specific and I wanted to get everything working first.

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+import scala.concurrent.ExecutionContext
+import scala.sys.process.Process
+
+/**
+  * This integration test starts older Marathon versions one after another and finishes this upgrade procedure with the
+  * current build. In each step we verfiy that all apps are still up and running.
+  */
+@IntegrationTest
+class UpgradeIntegrationTest extends AkkaIntegrationTest with MesosClusterTest with ZookeeperServerTest with Eventually {
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: I will mark the test to only run in the CI.

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      Then("All apps from 1.4.9 are still running")
+      val tasksAfterUpgradeToCurrent = marathonCurrent.client.tasks(app_146.id.toPath).value
+      tasksAfterUpgradeToCurrent should be(tasksBeforeUpgrade)
+
+      marathonCurrent.close()
+      //TODO(karsten): We leak processes.
+    }
+  }
+
+  // TODO(karsten): I'd love to have this test extend MarathonTest but I get NullPointerException for Akka.
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: Any help is welcome. I suspect that Akka is created lazily. Thus we get a NullPointerException when we try to access it at the wrong moment.

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+      When("Marathon is upgraded to the current version")
+      marathonCurrent.start().futureValue
+      (marathonCurrent.client.info.entityJson \ "version").as[String] should be("1.6.0-SNAPSHOT")
+
+      Then("All apps from 1.4.9 are still running")
+      val tasksAfterUpgradeToCurrent = marathonCurrent.client.tasks(app_146.id.toPath).value
+      tasksAfterUpgradeToCurrent should be(tasksBeforeUpgrade)
+
+      marathonCurrent.close()
+      //TODO(karsten): We leak processes.
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: `MesosTest.teardown` should work but we do not wait long enough for all tasks to be killed.

**Comment by jeschkies** [comment]
I will add persistent apps and pods in two separate follow ups.

File: src/main/scala/mesosphere/marathon/io/IO.scala [hunk 5]
```diff
@@ -1,7 +1,7 @@
 package mesosphere.marathon
 package io
 
-import java.io.{ Closeable, File, FileNotFoundException, InputStream, OutputStream }
+import java.io._
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/io/IO.scala]: Wow, wow, wow. :)

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/io/IO.scala]: 🙈 it was IntelliJ. I'm sorry.

File: src/main/scala/mesosphere/marathon/io/IO.scala [hunk 6]
```diff
@@ -1,10 +1,21 @@
 package mesosphere.marathon
 package io
 
-import java.io.{ Closeable, File, FileNotFoundException, InputStream, OutputStream }
+import java.io.{
```
Comment (zen-dog) [inline src/main/scala/mesosphere/marathon/io/IO.scala]: One import per string? O_O wow, wow...

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+/**
+  * This integration test starts older Marathon versions one after another and finishes this upgrade procedure with the
+  * current build. In each step we verfiy that all apps are still up and running.
+  */
+@IntegrationTest
+class UpgradeIntegrationTest extends AkkaIntegrationTest with MesosClusterTest with ZookeeperServerTest with Eventually {
+
+  // Integration tests using docker image provisioning with the Mesos containerizer need to be
+  // run as root in a Linux environment. They have to be explicitly enabled through an env variable.
+  val envVar = "RUN_MESOS_INTEGRATION_TESTS"
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: I suggest moving this envVar in one of the basic IT test objects. all tests should use the same ;)

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  case class Marathon149(suiteName: String, masterUrl: String, zkUrl: String)(
+      implicit
+      val system: ActorSystem, val mat: Materializer, val ctx: ExecutionContext, val scheduler: Scheduler) extends BaseMarathon {
+
+    val marathon149Package = Files.createTempDirectory("marathon-1.4.9").toFile
+    marathon149Package.deleteOnExit()
+
+    def downloadAndExtract() = {
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: This method should be extracted and generalized for all 3 marathons

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+      FileUtils.copyURLToFile(new URL("https://downloads.mesosphere.com/marathon/releases/1.4.9/marathon-1.4.9.tgz"), tarball)
+      IO.extractTGZip(tarball, marathon149Package)
+    }
+
+    override val processBuilder = {
+      val java = sys.props.get("java.home").fold("java")(_ + "/bin/java")
+      val jar = new File(marathon149Package, "marathon-1.4.9/target/scala-2.11/marathon-assembly-1.4.9.jar").getCanonicalPath
+      val cmd = Seq(java, "-Xmx1024m", "-Xms256m", "-XX:+UseConcMarkSweepGC", "-XX:ConcGCThreads=2",
+        // lower the memory pressure by limiting threads.
+        "-Dakka.actor.default-dispatcher.fork-join-executor.parallelism-min=2",
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: ...and these general memory setting too ;)

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        val tasks = marathon156.client.tasks(app_156.id.toPath).value
+        tasks should have size (1)
+        tasks.foreach { task =>
+          task.state should be("TASK_RUNNING")
+        }
+        tasks
+      }
+
+      And("All apps from 1.4.9 are still running")
+      marathon156.client.tasks(app_149.id.toPath).value should be(tasksBeforeUpgrade)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: even with 1 task it's better to take `should contain theSameElementsAs `

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+      marathonCurrent.client.tasks(app_156_fail.id.toPath).value should not be (originalApp156FailedTasks)
+
+      marathonCurrent.close()
+    }
+  }
+
+  def killTask(appName: String): Unit = {
+    val pidPattern = """([^\s]+)\s+([^\s]+)\s+.*""".r
+    val pids = Process("ps aux").!!.split("\n").filter { process =>
+      process.contains("src/test/python/app_mock.py") && process.contains(appName)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: What do you think about giving every app of this test an *additional* UUID in name and just checking `process.contains(uniqueAppName)` here?

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+      And(s"App ${app_149_fail.id} fails")
+      killTask("app-149-fail")
+
+      // Pass upgrade to 1.5.6
+      And("Marathon is upgraded to 1.5.6")
+      marathon156.start().futureValue
+      (marathon156.client.info.entityJson \ "version").as[String] should be("1.5.6")
+
+      And("new apps in Marathon 1.5.6 are added")
+      val app_156 = appProxy(testBasePath / "app-156", "v1", instances = 1, healthCheck = None)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: Please give them apps longer and more expressive names e.g. `migration-app-156-with-persistent-volumes`. Makes it much easier later to debug and reduces the chance of app name collision.

File: src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+      Then("All apps from 1.4.9 and 1.5.6 are still running")
+      marathonCurrent.client.tasks(app_149.id.toPath).value should be(originalApp149Tasks)
+      marathonCurrent.client.tasks(app_156.id.toPath).value should be(originalApp156Tasks)
+
+      And("All apps from 1.4.9 and 1.5.6 are recovered and running again")
+      eventually { marathonCurrent should have(runningTasksFor(app_149_fail.id.toPath, 1)) }
+      marathonCurrent.client.tasks(app_149_fail.id.toPath).value should not be (originalApp149FailedTasks)
+
+      eventually { marathonCurrent should have(runningTasksFor(app_156_fail.id.toPath, 1)) }
+      marathonCurrent.client.tasks(app_156_fail.id.toPath).value should not be (originalApp156FailedTasks)
```
Comment (zen-dog) [inline src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala]: here and above: `should contain theSameElementsAs`!

**Pull request merged**